### PR TITLE
Enhancement/setup backend

### PR DESCRIPTION
**Pull request by matty22**
Basic backend setup with dB, node server, and some basic Express routes, package.json and some other configuration files setup.

### CODE REVIEW

**Comment by matty22** [comment]
Closes #28

**Comment by matty22** [comment]
A question I have on this setup for anyone who reviews:

In bin/configuredb.sh the credentials for logging into the postgres server as a superadmin are there in plaintext. That file needs to be in the repo in case anyone needs to run `yarn run resetdb` but I feel like having those credentials there is dangerous.

Maybe I can fetch them out of a file I have in gitignore and insert the username and password that way?

File: bin/www [hunk 1]
```diff
@@ -0,0 +1,8 @@
+#!/usr/bin/env node
+
+const app = require('../app');
+const port = 3000;
```
Comment (lacyjpr) [inline bin/www]: I think we need a different port here, CRA defaults to 3000

File: db/index.js [hunk 2]
```diff
@@ -0,0 +1,11 @@
+const { Pool } = require('pg');
+const {
+  user,
+  host,
+  database,
+  password,
+  port,
+} = require('../secrets/db_configuration');
```
Comment (lacyjpr) [inline db/index.js]: Not sure if this will work on Heroku: since the secrets directory isn't under source control Heroku won't know about it

Comment (lacyjpr) [reply db/index.js]: Is there a way to use .env variables instead?

File: bin/configuredb.sh [hunk 3]
```diff
@@ -0,0 +1,13 @@
+#!/bin/bash
+
+export PGPASSWORD="L!llyP0nd"
+
```
Comment (lacyjpr) [inline bin/configuredb.sh]: To answer your question, I think a .env variable would work here

Comment (lacyjpr) [reply bin/configuredb.sh]: https://stackoverflow.com/questions/22312671/node-js-setting-environment-variables

File: bin/www (hunk 1 above)
Comment (matty22) [reply bin/www]: Any foreseeable issue if I use 4200?

Comment (lacyjpr) [reply bin/www]: I can't say for sure:) I've used 5000 in the past, I *think* any port that is not in use will work

File: bin/configuredb.sh (hunk 3 above)
Comment (matty22) [reply bin/configuredb.sh]: In that thread, someone mentions this package: https://github.com/motdotla/dotenv
Seems pretty straightforward approach to me. Eventually we may have to adjust them when we try to deploy to heroku, but for now it'll keep that stuff hidden from the repo

File: app.js [hunk 4]
```diff
@@ -0,0 +1,16 @@
+const express = require('express');
+const app = express();
```
Comment (Kornil) [inline app.js]: this is inconsistent and hard to read, `require` and THEN define stuff.

File: routes/users.js [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+router.get('/', (request, response) => {
+  pool.query('SELECT * FROM users ORDER BY id ASC', (err, res) => {
+    if (err) return next(err);
+    response.json(res.rows);
+  });
+});
+
+router.get('/:id', (request, response, next) => {
+  const { id } = request.params;
+  pool.query('SELECT * FROM users WHERE id = $1', [id], (err, res) => {
```
Comment (Kornil) [inline routes/users.js:14]: I have a bad feeling about this. The client interacts directly with database, not a good way to do things.

**Review by Kornil — COMMENTED** [review]
Add a linter, the lack of newlines is driving me insane.

**Comment by Kornil** [comment]
> In bin/configuredb.sh the credentials for logging into the postgres server as a superadmin are there in plaintext. That file needs to be in the repo in case anyone needs to run yarn run resetdb but I feel like having those credentials there is dangerous.

@matty22 You are correct, store as ENV variables and `rebase` so they get out of git history

File: routes/users.js (hunk 5 above)
Comment (matty22) [reply routes/users.js:14]: I don't follow. This is just like how a mongo query would be performed:

* Request comes to server from client
* Server queries database
* Server returns data to client

How else would this be handled?

Comment (Kornil) [reply routes/users.js:14]: create APIs that the clients interacts with instead of querying the db

Comment (matty22) [reply routes/users.js:14]: Is this not the API? The Express routes? The `pool.query` is how the `pg` node package queries the postgres database. I assume it works similarly to `.findById` in Mongo.

I should just remove these queries for now anyways. They're just example filler, but I am still confused on your comment and be not confused would be ideal. 😄

Comment (Kornil) [reply routes/users.js:14]: http://mherman.org/blog/2015/02/12/postgresql-and-nodejs/#server-side-routes you create an API route, client interacts with the predefined API route

Comment (matty22) [reply routes/users.js:14]: Here's the type of query I'm using in the examples in this file: https://node-postgres.com/features/pooling#single-query. It even says:

> This is the preferred way to query with node-postgres if you can as it removes the risk of leaking a client.

But I'm still learning how to use node-postgres and integrating it with some Express routes, so it's entirely possible I'm misunderstanding it all.

**Comment by lacyjpr** [comment]
Live here: https://media-ycmretxgso.now.sh
Navigate to https://media-ycmretxgso.now.sh/users to verify it works

**Pull request merged**